ln: reduce size_of Event from 1856 B -> 720 B, prep for inline storage in offer `*Contents`

phlip9 opened the pull request:
It looks like the `Event` enum has gotten pretty large @ 1856 B, which forces even small variants like `Event::OnionMessagePeerConnected { peer_node_id: PublicKey }` to waste a bunch of memory. It also blows up the size of our handler Future(s), since we move `Event`s into them.

Fortunately we can clean up some of the low-hanging fruit pretty easily. Here's two diffs that reduce `size_of::<Event>()` from ~~1680 B -> 576 B~~ 1856 B -> 720 B.

1. Box `InvoiceContents` in `Bolt12Invoice` and `StaticInvoice`. This shrinks `Event` from 1856 B -> 1072 B
  + `InvoiceContents` is private, so this shouldn't be a breaking change
2. Box `AnchorDescriptor` in `BumpTransactionEvent`. This shrinks `Event` from 1072 B -> 720 B.
  + `AnchorDescriptor` is technically public and boxing it is a semver breaking change, but the struct's pretty deep in there... Guess I'll leave that to your discretion.

We could go even further to 320 B by boxing `PaymentPurpose`, but that feels like a much more invasive / semver breaking change.

tnull reviewed (commented):
To be honest, I'm not sure if its preferable to trade the size against an increased risk of heap fragmentation? Did you consider that, and how did you decide a smaller `Event` type was worth it?

vincenzopalazzo reviewed (commented):
I’m not sure I understand the reason for this change. Did you encounter any issues, or is it just for size reduction?

wpaulino reviewed (commented):
I can see how the large size can be a constraint for environments with smaller stack sizes/limited memory. Events are tracked in LDK within a Vec, so the memory of each one already lives on the heap. The issue comes from moving each event into the event handler, so the 1680B will be put on the stack every time regardless of the event variant size. Perhaps we should modify the event handler to take events by reference instead?

phlip9 commented:
> To be honest, I'm not sure if its preferable to trade the size against an increased risk of heap fragmentation? Did you consider that, and how did you decide a smaller Event type was worth it?

Right, there is slight heap fragmentation in 1.5 / 25 variants in exchange for reduced heap+stack usage, cache misses, etc for all other variants.

I'm not super familiar with the bench setup in LDK, but I tried running `channelmanager::bench_sends` before and after this PR on an M1 mac:

```
dev/ldk/bench$ RUSTFLAGS="--cfg=ldk_bench" c bench
   Compiling lightning v0.2.0+git (dev/ldk/lightning)
   Compiling lightning-rapid-gossip-sync v0.2.0+git (dev/ldk/lightning-rapid-gossip-sync)
   Compiling lightning-persister v0.2.0+git (dev/ldk/lightning-persister)
   Compiling lightning-bench v0.0.1 (dev/ldk/bench)
    Finished `bench` profile [optimized + debuginfo] target(s) in 3m 18s
     Running benches/bench.rs (target/release/deps/bench-126259c02b665b62)
bench_sends             time:   [4.3944 ms 4.4082 ms 4.4229 ms]
                        change: [-1.5040% -0.7683% -0.1187%] (p = 0.03 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild
```

Within the noise threshold, but not nothing. An OnionMessenger process_events bench would probably make this more apparent.

Ultimately the most performant approach IMO would be to just hand a batch of events to the consumer instead of the current "event-by-event" handling. That would complicate the error handling, but would let us persist batches of events that we need to handle asynchronously.

phlip9 commented:
Sorry, to add on to this, we currently have some layers of async handlers for each `Event` that pass by value. Like:

```rust
async fn event_handler.get_ldk_handler_future(event) // 14624 B (!)
async fn event_handler.handle_inline(event) // 12928 B
async fn event_handler.handle_event(id, event) // 11136 B
async fn event_handler.persisant_and_spawn_handler(event) // 1872 B
async fn event::handle_payment_claimable(id, {..}) // ...
...
```

And each layer gets blown up by + size_of Event. So reducing size_of Event would have a multiplicative effect in reducing the size of our futures.

phlip9 committed:
offers: box `*Contents`. reduce size_of `Event` 1856 -> 1072

Prep for inline storage for `description`, `issuer`, `payer_note`

* invoice: box `InvoiceContents`
  reduces `mem::size_of::<Event>()` : 1856 -> 1072
* invoice_request: box contents
* offer: box `OfferContents`
* refund: box `RefundContents`
* static invoice: box `InvoiceContents`

phlip9 committed:
events: box `AnchorDescriptor`. reduce size_of `Event` 1072 -> 720
